src/core/lib/gpr/tls.h

```diff
@@ -67,13 +61,45 @@ namespace grpc_core {
 template <typename T>
 class PthreadTlsImpl : TlsTypeConstrainer<T> {
  public:
-  PthreadTlsImpl() = default;
   PthreadTlsImpl(const PthreadTlsImpl&) = delete;
   PthreadTlsImpl& operator=(const PthreadTlsImpl&) = delete;
 
-  void Init() { GPR_ASSERT(0 == pthread_key_create(&key_, nullptr)); }
-
-  void Destroy() { GPR_ASSERT(0 == pthread_key_delete(key_)); }
+  // Achtung! This class emulates C++ `thread_local` using pthread keys. Each
+  // instance of this class is a stand in for a C++ `thread_local`. Think of
+  // each `thread_local` as a *global* pthread_key_t and a type tag. An
+  // important consequence of this is that the lifetime of a `pthread_key_t`
+  // is precisely the lifetime of an instance of this class. To understand why
+  // this is, consider the following scenario given a fictional implementation
+  // of this class which creates and destroys its `pthread_key_t` each time
+  // a given block of code runs (all actions take place on a single thread):
+  //
+  // - instance 1 (type tag = T*) is initialized, is assigned `pthread_key_t` 1
+  // - instance 2 (type tag = int) is initialized, is assigned `pthread_key_t` 2
+  // - instances 1 and 2 store and retrieve values; all is well
+  // - instances 1 and 2 are de-initialized; their keys are released to the pool
+  //
+  // - another run commences
+  // - instance 1 receives key 2
+  // - a value is read from instance 1, it observes a value of type int, but
```

Comment:
Setting the value is per-thread. So it would work when there's just one thread, but not in the general case.

Reply:
It also feels odd to do that _here_, since it essentially clobbers an unknown previous value. It would produce behavior that is different from the native `thread_local` behavior.

Reply:
I see.

Another option to avoid the style guide violation would be to change the global variable from `PthreadTlsImpl<T>` to `PthreadTlsImpl<T>*`.  Then in the places where we were calling `gpr_tls_init()`, we can instead call `new PthreadTlsImpl<T>`, and in the places where we were calling `gpr_tls_destroy()`, we can instead call `delete`.  That way, the `PthreadTlsImpl<T>` itself is not a global variable, only the pointer to it is.

Reply:
That has the same problem as the situation today. Each instance of this class must exactly model a `thread_local`. In your proposal, the pointer would at different times have different pointees, which might be assigned a different `pthread_key_t` and thus once again lead to undefined behavior.

You could solve the problem by never calling `pthread_key_delete`, but that has problems as well.